src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -499,14 +521,22 @@ bool ClusterTaskManager::CancelTask(const TaskID &task_id) {
        shapes_it++) {
     auto &work_queue = shapes_it->second;
     for (auto work_it = work_queue.begin(); work_it != work_queue.end(); work_it++) {
-      const auto &task = std::get<0>(*work_it);
+      const auto &task = (*work_it)->task;
       if (task.GetTaskSpecification().TaskId() == task_id) {
+        if ((*work_it)->dispatch_finished) {
+          RAY_LOG(INFO) << "Canceling Task  " << task_id
+                        << " failed because task already dispatched.";
+          work_queue.erase(work_it);
+          return false;
+        }
         RemoveFromBacklogTracker(task);
+        RAY_LOG(DEBUG) << "Canceling task " << task_id << " from dispatch queue.";
         ReplyCancelled(*work_it);
         if (!task.GetTaskSpecification().GetDependencies().empty()) {
           task_dependency_manager_.RemoveTaskDependencies(
               task.GetTaskSpecification().TaskId());
         }
+        (*work_it)->canceled = true;
```

Comment:
Can you add a comment to explain when the task actually gets erased?

Reply:
@stephanie-wang I have erased this task in next line? https://github.com/ray-project/ray/blob/7378327c59af9bf48291a17eaee1b43c6ce832d0/src/ray/raylet/scheduling/cluster_task_manager.cc#L540

Reply:
Ah yes, I meant to explain when *work_it gets erased. I think that happens when the dispatch loop gets triggered again, right?

Reply:
We have two place to erase task now: `CancelTask` and `erase_from_dispatch_queue_fn`.